### ISSUE TITLE
[BUG]  widget button actions delayed after approx 900 tick_callback

### ISSUE THREAD

**Issue by pannet1** [bokeh/bokeh#12141]
### Software versions

Python version      :  3.8.13 (default, Mar 17 2022, 00:14:49) 
IPython version     :  8.3.0
Tornado version     :  6.1
Bokeh version       :  2.4.2
BokehJS static path :  /home/pannet1/Programs/38py/lib/python3.8/site-packages/bokeh/server/static
node.js version     :  v16.15.0
npm version         :  8.5.5
Operating system    :  Linux-5.17.9-arch1-1-x86_64-with-glibc2.34

### Expected behavior

no response for button clicks after 900 runs approximately

### Observed behavior

delayed response for button clicks. delay increases over time.

### Example code

```Python
# bokeh graphics
doc = curdoc()
TOOLS = "pan,wheel_zoom,box_zoom,reset"
source = ColumnDataSource(
    dict(time=[], open=[], close=[])
)
span_atm = Span(location=0,dimension='width', line_dash='dashed', line_width=3) 
fig_line = figure(
        tools=TOOLS,
        x_axis_type = 'datetime',
        y_axis_location = "right",
        tooltips=[('time', '@time{%F %H:%M}'), 
                 ('open', '@open{0}'),
                 ('close', '@close{0}')])

fig_line.hover.formatters = {'@time': 'datetime'}
fig_line.xaxis.major_label_orientation = pi / 4
fig_line.line(x='time',
          y='close',
          line_width=1,
          source=source)

fig_line.add_layout(span_atm)

# init widgets left out for brevity

def place_order(option, side)
   #do some other blocking task
   pass


def widget_events():
    btn_c_buy.on_click(partial(place_order, option="CE", side="BUY"))
    btn_c_sel.on_click(partial(place_order, option="CE", side="SELL"))
    btn_p_sel.on_click(partial(place_order, option="PE", side="SELL"))
    btn_p_buy.on_click(partial(place_order, option="PE", side="BUY"))

widget_events()

async def update(dct_ul:dict, dct_ltp:dict, dct_circle:dict, ttl:int): 
    global OPEN_PRICE
    global counter
    logger.info(f"{counter} update called")

    source.stream( dict( time=[ dct_ul['time'] ], open =[ OPEN_PRICE ], close=[ dct_ul['close'] ]) )
    # show price as bokeh title
    fig_line.title.text = ulying_sym + " / " + str( OPEN_PRICE  )
    fig_line.title.text_color = clr_drk_bear if OPEN_PRICE > dct_ul['close'] else clr_drk_bull 

    txt_pnl.value = str(ttl)   
    span_atm.location = dct_ul['atm']
    # show rounded atm strike as text
    txt_atm.value = str(dct_ul['atm_100']) 

    if dct_ul.__contains__('clr'):
        fig_line.circle( x = [ dct_ul['time']],  y=[ dct_ul['atm'] ], color= [ dct_ul['clr']], size=14)

    fig_line.circle(x='x' , y='y', source=dct_circle)    

    OPEN_PRICE = dct_ul['close']
    logger.info(f'{counter } update complete')
    #update function ends here

HIGH = 0
counter = 0
def blocking_task():
    while True:
        global EXCSYM
        global HIGH
        global counter

        dct_ltp = make_http_request (EXCSYM)

        counter += 1
        logger.info(f'{ counter } blocking sleep for { MS_TAKEN } seconds')

        sleep(8)
        dct_ul = dct_ltp[ulying_sym]
        # pop underlying from the main ltp dict
        dct_ltp.pop(ulying_sym)
        
        # atm gui
        dct_ul['atm'] =  round(dct_ul['close'] / strike_step) * strike_step 
        new_step = strike_step * 2
        if dct_ul['close'] > HIGH:
            HIGH = dct_ul['close']
        dct_ul['atm_100'] =  round( HIGH / new_step ) * new_step

        # milestones
        if int(dct_ul['close']) < int(dct_ul['atm']) < int(OPEN_PRICE):
            dct_ul['clr'] = clr_drk_bear
        elif int(OPEN_PRICE) < int(dct_ul['atm']) < int(dct_ul['close']): 
            dct_ul['clr'] = clr_drk_bull

        # find mtm only if order button is not pressed      
        dct_m2m = {}
        x = []
        y = []
        for k,v in dct_ltp.items():
            #flatten list
            dct_m2m[k] = v['close']
            # draw ltp circles
            ce = re.findall(r"(\d+)CE", k) 
            pe = re.findall(r"(\d+)PE", k) 
            if ce:
                real_val = int(ce[0]) + v['close']
                x.append(v['time'])
                y.append(real_val)
            elif pe:
                real_val = int(pe[0]) - v['close']
                x.append(v['time'])
                y.append(real_val)

        dct_circle = { 'x': x, 'y': y} 
    
        total_mtm = int(com.total_mtm)
        if max_mtm_loss > total_mtm:
            logger.info(f"max loss { total_mtm } reached")

        logger.info(f"{ counter } blocking task completed")  
        doc.add_next_tick_callback(partial(update, dct_ul=dct_ul, dct_ltp=dct_ltp, dct_circle=dct_circle, ttl=total_mtm))

doc.add_root(row(fig_line, widget))
thread = Thread(target=blocking_task)
thread.start()
```
```


### Stack traceback or browser console output

_No response_

### Screenshots

_No response_

**Comment by bryevdv** [comment]
@pannet1 We will need more information from you before this can be invegstigate:

* The MRE is not complete, it is missing at least imports. The MRE should be something that can be copy and pasted into a new file and run immediately, without any ambiguity, speculation, or "gaps" to fill in. These are a huge potential source of wasted time and effort. 

* You mention "no response for button clicks after 900 runs approximately" however the callbacks above are complete no-ops, so there would be no way to see them doing or not doing anything. You need to fill in concrete code that you know for certain will reproduce the issue / what you want us to see, clearly and obviously.

**Comment by pannet1** [comment]
The actual code needs credentials for making API requests. I will try to make a workaround to create the MRE. Simple guess is to wrap a sleep() for the button clicks, hoping to introduce the racing like condition for the blocking task and the update function.

**Comment by pannet1** [comment]
try clicking any of the buttons on the right in tandem. even if you let it catchup after sometime, you could see the button start loosing the sensitivity to click completely. also at this time when you see the logger on the screen, there seems to be a race condition.

```
import sys
from threading import Thread

from time import sleep
import pendulum
from functools import partial
from math import pi
import re

from bokeh.layouts import column, row
from bokeh.models import ColumnDataSource, Span
from bokeh.models.widgets import (Button, Div, RadioButtonGroup, Select,
                                  TextInput)
from bokeh.plotting import curdoc, figure

from logzero import logger
import json


# constants
max_mtm_loss = -2500
default_qty = "50"
underlying  = { 'exc': 'NFO', 'pfx': ':', 'sym': 'NIFTY', 'sfx': '22MAYFUT'   }
ulying_sym  = underlying['sym'] + underlying['sfx']
expires     = ["22MAY","22JUN"]
strike_step = 50
MS_TAKEN    = 1  # 1k ms = 1sec

clr_drk_bull= "#115600"
clr_drk_bear= "firebrick"

# mutated globals
OPEN_PRICE  = 0.0
IS_DIRTY    = True


# bokeh graphics
doc = curdoc()
TOOLS = "pan,wheel_zoom,box_zoom,reset"
source = ColumnDataSource(
    dict(time=[], open=[], close=[])
)
span_atm = Span(location=0,dimension='width', line_dash='dashed', line_width=3)
fig_line = figure(
        tools=TOOLS,
        x_axis_type = 'datetime',
        y_axis_location = "right",
        tooltips=[('time', '@time{%F %H:%M}'),
                 ('open', '@open{0}'),
                 ('close', '@close{0}')])

fig_line.hover.formatters = {'@time': 'datetime'}
fig_line.xaxis.major_label_orientation = pi / 4
fig_line.line(x='time',
          y='close',
          line_width=1,
          source=source)

fig_line.add_layout(span_atm)

fig_line.title.align = "center"
fig_line.title.text_font_size = "24px"

# trade panel
lbl_title = Div(text="""
    <h2 style='text-align:center'>Trade Panel</h2>
    """,
            height=50,
            width=200)
btn_c_buy = Button(label="Buy CE",
               button_type="success",
               sizing_mode="fixed",
               height=30,
               width=100)
txt_c_buy = TextInput(value="1000", height=30, width=90)
row_c_buy = row(btn_c_buy, txt_c_buy)
btn_c_cvr = Button(label="Cover",
               button_type="warning",
               sizing_mode="fixed",
               height=30,
               width=60)
btn_c_sel = Button(label="Sell",
               button_type="danger",
               sizing_mode="fixed",
               height=30,
               width=60)
txt_c_sel = TextInput(value="300", height=30, width=60)
row_c_sel = row(btn_c_cvr, btn_c_sel, txt_c_sel)

sel_expiry = Select(value=expires[0],
                    options=expires,
                    sizing_mode="fixed",
                    height=30,
                    width=130)
txt_qty = TextInput(value="50", height=30, width=60)
row_mid = row(sel_expiry, txt_qty)

btn_p_cvr = Button(label="Cover",
                   button_type="warning",
                   sizing_mode="fixed",
                   height=30,
                   width=60)
btn_p_sel = Button(label="Sell",
                   button_type="danger",
                   sizing_mode="fixed",
                   height=30,
                   width=60)
txt_p_sel = TextInput(value="-300", height=30, width=60)
txt_atm   = TextInput(value="0", height=30, width=190)
row_p_sel = row(btn_p_cvr, btn_p_sel, txt_p_sel)
btn_p_buy = Button(label="Buy PE",
                   button_type="success",
                   sizing_mode="fixed",
                   height=30,
                   width=100)
txt_p_buy = TextInput(value="-1000", height=30, width=90)
row_p_buy = row(btn_p_buy, txt_p_buy)
lbl_pnl = Div(text="""
    <h4>Marked to Market</h4>
    """,
              height=30,
              width=200)
txt_pnl = TextInput(value="0", sizing_mode="fixed", height=30, width=200)
widget = column(lbl_title, row_c_buy, row_c_sel, row_mid, txt_atm, row_p_sel, row_p_buy,
                lbl_pnl, txt_pnl)

def get_ltps(excsym):
    dct_ltp = {}
    resp = kite.ltp(excsym)
    for k,v in resp.items():
        exc_col_sym = k.split(":")
        dct_ltp[exc_col_sym[1]] = { "time": pendulum.now(), "close": v['last_price'] }
    return dct_ltp

'''
 order functions
'''
def place_order(option, side):
    logger.info(f'button pressed with option {option} and side {side}' )
    if option=="CE":
        away = txt_c_buy.value if side=="BUY" else txt_c_sel.value
    elif option=="PE":
        away = txt_p_buy.value if side=="BUY" else txt_p_sel.value

    strike = int(txt_atm.value) + int(away)
    symbol = underlying['sym'] + sel_expiry.value + str(strike) + option

    order_time = pendulum.now()
    clr = clr_drk_bull if side == 'BUY' else clr_drk_bear
    real_val = 16000 + strike if option=='CE' else strike - 16000
    text = str(txt_qty.value) + " X " + str(strike)  + " @ "
    # draw order line_dash
    fig_line.ray(x=[ order_time ], y=[ real_val ], color=[clr] )
    fig_line.text(x=[ order_time], y=[ real_val], text=[ text ], y_offset=-2,
        text_color=[ clr ], text_font_size="14px")
    sleep(MS_TAKEN)


def widget_events():
    btn_c_buy.on_click(partial(place_order, option="CE", side="BUY"))
    btn_c_sel.on_click(partial(place_order, option="CE", side="SELL"))
    btn_c_cvr.on_click(partial(place_order, option="CE", side="SELL"))
    btn_p_sel.on_click(partial(place_order, option="PE", side="SELL"))
    btn_p_buy.on_click(partial(place_order, option="PE", side="BUY"))
    btn_p_cvr.on_click(partial(place_order, option="PE", side="BUY"))

widget_events()

async def update(dct_ul:dict, dct_ltp:dict, dct_circle:dict, ttl:int):
    global OPEN_PRICE
    global counter
    logger.info(f"{counter} update called")

    source.stream( dict( time=[ dct_ul['time'] ], open =[ OPEN_PRICE ], close=[ dct_ul['close'] ]) )
    # show price as bokeh title
    fig_line.title.text = ulying_sym + " / " + str( OPEN_PRICE  )
    fig_line.title.text_color = clr_drk_bear if OPEN_PRICE > dct_ul['close'] else clr_drk_bull

    txt_pnl.value = str(ttl)
    span_atm.location = dct_ul['atm']
    # show rounded atm strike as text
    txt_atm.value = str(dct_ul['atm_100'])

    if dct_ul.__contains__('clr'):
        fig_line.circle( x = [ dct_ul['time']],  y=[ dct_ul['atm'] ], color= [ dct_ul['clr']], size=14)

    fig_line.circle(x='x' , y='y', source=dct_circle)

    OPEN_PRICE = dct_ul['close']
    logger.info(f'{counter } update complete')
    #update function ends here

HIGH = 0
counter = 0
def blocking_task():
    while True:
        global EXCSYM
        global HIGH
        global counter

        counter += 1
        logger.info(f'{ counter } blocking sleep for { MS_TAKEN } seconds')

        sleep(MS_TAKEN)
        dct_ltp = { ulying_sym : { 'close': 16000 + counter, 'time': pendulum.now() },
                   'NIFTY22MAY16000CE' : { 'close': 200, 'time': pendulum.now()}
                   }
        dct_ul = dct_ltp[ulying_sym]
        # pop underlying from the main ltp dict
        dct_ltp.pop(ulying_sym)

        # atm gui
        dct_ul['atm'] =  round(dct_ul['close'] / strike_step) * strike_step
        new_step = strike_step * 2
        if dct_ul['close'] > HIGH:
            HIGH = dct_ul['close']
        dct_ul['atm_100'] =  round( HIGH / new_step ) * new_step

        # milestones
        if int(dct_ul['close']) < int(dct_ul['atm']) < int(OPEN_PRICE):
            dct_ul['clr'] = clr_drk_bear
        elif int(OPEN_PRICE) < int(dct_ul['atm']) < int(dct_ul['close']):
            dct_ul['clr'] = clr_drk_bull

        dct_m2m = {}
        x = []
        y = []
        for k,v in dct_ltp.items():
            #flatten list
            dct_m2m[k] = v['close']
            # draw ltp circles
            ce = re.findall(r"(\d+)CE", k)
            pe = re.findall(r"(\d+)PE", k)
            if ce:
                real_val = int(ce[0]) + v['close']
                x.append(v['time'])
                y.append(real_val)
            elif pe:
                real_val = int(pe[0]) - v['close']
                x.append(v['time'])
                y.append(real_val)

        dct_circle = { 'x': x, 'y': y}
        logger.info(f"{ counter } blocking task completed")
        doc.add_next_tick_callback(partial(update, dct_ul=dct_ul, dct_ltp=dct_ltp, dct_circle=dct_circle, ttl=5000))

doc.add_root(row(fig_line, widget))
thread = Thread(target=blocking_task)
thread.start()
```

**Comment by bryevdv** [comment]
You are adding entire new glyphs every second:

```
    if dct_ul.__contains__('clr'):
        fig_line.circle( x = [ dct_ul['time']],  y=[ dct_ul['atm'] ], color= [ dct_ul['clr']], size=14)

    fig_line.circle(x='x' , y='y', source=dct_circle)
``` 
This is absolutely *not* the way to use Bokeh. Typical Bokeh usage is to add a few glyphs up front to define the "shape" of the plot, and vectorize those glyphs over the data in CDS. Bokeh is can be very efficient at updating CDSes. Glyphs are heavyweight objects so adding them dynamically is expensive, and not only that, you are adding an ***unbounded*** number of glyphs. This is a memory leak in your code and will absolutely fall over eventually. 

Additionally for the data you do `stream` you do not specify any `rollover` so that will also accumulate infinitely so that will also eventually be a problem at some point. 

TLDR; Bokeh best practice is to add a fixed set of glyphs up front then only update their CDS data sources dynamically. 

I am not 100%  sure if these are all of the problems in your code. But these are definitely  usage issues, and I don't have time to look into it further. Since you have already opened a separate topic on the Discourse, I am going to close this for now to discuss any further developments or usage questions there for now.
